contributors/design-proposals/storage/container-storage-interface-cluster-registry.md

```diff
@@ -0,0 +1,158 @@
+
+# Summary
+
+This document proposes a "CSI cluster registration mechanism".
+
+This mechanism will enable:
+* A CSI volume driver to register itself with Kubernetes when it is deployed.
+* A CSI volume driver to customize how Kubernetes interacts with it (e.g. skip attach process because the driver doesn't support ControllerPublish, etc.).
+* A user or cluster admin to easily discover which CSI volume drivers are deployed on their Kubernetes cluster.
+* Be optional -- a driver may choose not to use it (and will get the default set of behaviors).
+
+
+## Background & Motivations
+
+Kubernetes supports the Container Storage Interface (CSI) to enable third party storage developers to deploy volume drivers exposing new storage systems in Kubernetes without having to touch the core Kubernetes code.
+Support for CSI was introduced as alpha in Kubernetes v1.9, and moved to beta in v1.10.
+See "CSI Volume Plugins in Kubernetes Design Doc" in the "Links" section below for details on how an arbitrary CSI volume driver is deployed and interacts with Kubernetes.
+The beta implementation of CSI has a number of limitations, including:
+* CSI drivers must be deployed with the provided CSI external-attacher sidecar container, even if they don’t implement ControllerPublishVolume.
+  * Meaning a CSI volume driver deployed on kubernetes has to deploy a CSI `external-attacher` container even if the volume driver doesn't require a "volume attach" operation.
+  * In this case, the CSI `external-attacher` container basically does a no-op in response to Kubernetes `VolumeAttachment` objects to allow Kubernetes to continue with the mounting process.
+* Users and cluster admins have no easy way to discover what CSI volume drivers are deployed on their Kubernetes cluster.
+
+The proposed "CSI cluster registration mechanism" should address these issues.
+
+### Links
+
+* [Container Storage Interface (CSI) Spec](https://github.com/container-storage-interface/spec/blob/master/spec.md)
+* [CSI Volume Plugins in Kubernetes Design Doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/container-storage-interface.md)
+
+## Objective
+
+The objective of this document is to document all the requirements for enabling a cluster level registry for a CSI compliant volume drivers in Kubernetes.
+
+## Goals
+
+* Allow a CSI volume driver (that opts-in to new mechanism) to configure how Kubernetes should interact with it.
+* Improve discoverability by users/cluster admins of a CSI volume driver (that opts-in to new mechanism) deployed on the Kubernetes cluster.
+
+## Non-Goals
+
+* Define how CSI volume driver should be deployed on Kubernetes.
+* Require use of new CSI cluster registry mechanism for all CSI volume drivers deployed on Kubernetes (the mechanism will be optional/opt-in).
+
+## Design Overview
+
+A new custom resource will be automatically be installed on Kubernetes clusters.
```

Comment:
a new custom resource **definition**? CR is optional.

Reply:
Fixed